extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -333,18 +348,54 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
       dev().warn(TAG, `Frame already exists, sra: ${this.useSra}`);
       return '';
     }
+
+    const ampRtcPageElement = document.getElementById('amp-rtc');
+    let rtcRequestPromise = null;
+    let rtcConfig;
+    if (ampRtcPageElement) {
+      rtcConfig = tryParseJson(ampRtcPageElement.innerHTML);
+      if (rtcConfig && typeof rtcConfig == 'object'
+      && !!rtcConfig['doubleclick']) {
+        rtcRequestPromise = this.sendRtcRequestPromise_(/** @type {!Object} */(
+            rtcConfig['doubleclick']));
+      }
+    }
     // TODO(keithwrightbos): SRA blocks currently unnecessarily generate full
     // ad url.  This could be optimized however non-SRA ad url is required to
     // fallback to non-SRA if single block.
     this.populateAdUrlState();
     // TODO: Check for required and allowed parameters. Probably use
     // validateData, from 3p/3p/js, after noving it someplace common.
     const startTime = Date.now();
-    return getPageLevelParameters_(this.win, this.getAmpDoc(), startTime)
-        .then(pageLevelParameters =>
-        googleAdUrl(this, DOUBLECLICK_BASE_URL, startTime,
-            Object.assign(this.getBlockParameters_(), pageLevelParameters),
-          ['108809080']));
+    const pageLevelParametersPromise = getPageLevelParameters_(
+        this.win, this.getAmpDoc(), startTime);
+    const blockParameters = this.getBlockParameters_();
+
+    return Promise.all(
+      [pageLevelParametersPromise, rtcRequestPromise]).then(values => {
+        const pageLevelParameters = values[0];
+        const rtcResponse = values[1];
+        const parameters = Object.assign(blockParameters, pageLevelParameters);
+
+        if (rtcResponse) {
+          if (!!rtcResponse['targeting']) {
+            const targeting = deepMerge(this.jsonTargeting_['targeting'] || {},
+                rtcResponse['targeting'] || {});
+            const exclusions = deepMerge(
+                this.jsonTargeting_['categoryExclusions'] || {},
+                rtcResponse['exclusions'] || {});
+            parameters['scp'] = serializeTargeting_(targeting, exclusions);
+          } else {
+            // Default is to send request without RTC, so only reject if
+            // pub explicitly says not to.
+            if (rtcConfig['doubleclick']['sendAdRequestOnFailure'] === false) {
+              return Promise.reject();
```

Comment:
There are 3 failure conditions of which any failure should prevent ad request if indicated by pub:

- network failure
- timeout
- malformed response

This is addressing the last item but I would argue that only malformation we care about is failure to JSON parse which we can handle as part of sendRTCRequest flow